[X86][APX] Support peephole optimization with CCMP instruction

fzou1 opened the pull request:
This extends `opitimizeCompareInstr` to re-use previous CCMP results if the
previous comparison was with an immediates that was 1 bigger or smaller. Example:
```
CCMP x, 13, 2, 5
...
CCMP x, 12, 2, 5 ; can be removed if we change the SETg
SETg ...         ; x > 12 changed to SETge (x >= 13) & remove the 2nd
CCMP
```

fzou1 committed:
[X86] Support peephole optimization with CCMP instruction

This extends `opitimizeCompareInstr` to re-use previous CCMP results if
the
previous comparison was with an immediates that was 1 bigger or smaller.
Example:
```
CCMP x, 13, 2, 5
...
CCMP x, 12, 2, 5 ; can be removed if we change the SETg
SETg ...         ; x > 12 changed to SETge (x >= 13) & remove the 2nd
CCMP
```

KanRobert commented:
I think https://github.com/llvm/llvm-project/blob/main/llvm/test/CodeGen/X86/apx/ccmp.ll should reflect your change?

fzou1 committed:
Add checks and LIT tests

Check scc and dfv are same if FlagI is CCMP instruction and add more LIT
tests.

fzou1 commented:
> I think https://github.com/llvm/llvm-project/blob/main/llvm/test/CodeGen/X86/apx/ccmp.ll should reflect your change?

For now, there is no such pattern in that LIT test. I'll add some sub-tests to validate this optimization.

phoebewang commented on llvm/lib/Target/X86/X86InstrInfo.cpp:
```diff
@@ -4951,6 +4955,20 @@ bool X86InstrInfo::isRedundantFlagInstr(const MachineInstr &FlagI,
+    // The CCMP instruction should not be optimized if the scc/dfv in it is not
+    // same as the one in previous CCMP instruction.
+    unsigned Opcode = FlagI.getOpcode();
+    if (Opcode == X86::CCMP64ri32 || Opcode == X86::CCMP32ri ||
+        Opcode == X86::CCMP16ri || Opcode == X86::CCMP8ri)
```
Do we need `FlagI.getOpcode() == OI.getOpcode()`?

fzou1 replied:
Yes. Updated. It can prevent incorrect optimization (see CCMP + CMP combination in opt_redundant_flags_adjusted_imm_noopt_5 MIR sub-test). Thanks.

phoebewang replied:
This doesn't prevent  CCMP + CMP, just e.g., CCMP64ri32 + CCMP16ri. I don't know if we have this pattern in real workload.

fzou1 committed:
Add check to ensure CCMP opcode is same

fzou1 committed:
Add two sub-tests in ccmp LIT test for validate peephole optimization.

phoebewang reviewed (approved):
LGTM.

KanRobert commented on llvm/test/CodeGen/X86/apx/optimize-compare-ccmp.mir:
```diff
@@ -0,0 +1,396 @@
+    ; CHECK-NEXT: $bl = SETCCr 12, implicit $eflags
+    ; CHECK-NEXT: $bl = SETCCr 2, implicit $eflags
+    %0:gr64 = COPY $rsi
+    ; CCMP+SETCC   %0 != 42
+    CCMP64ri32 %0, 42, 2, 5, implicit-def $eflags, implicit $eflags
+    $cl = SETCCr 5, implicit $eflags
+    ; CCMP+SETCC   %0 >= 2; CCMP can be removed.
```
Comment here is incorrect?

fzou1 replied:
Nice catch. Updated it as well as the comments in llvm/test/CodeGen/X86/optimize-compare.mir test. Thanks.

fzou1 committed:
Update comments in test.

KanRobert reviewed (approved):
LGTM
